core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java

```diff
@@ -114,28 +113,21 @@ public QueryShardContext(QueryShardContext source) {
         this.types = source.getTypes();
     }
 
-    public void parseFieldMatcher(ParseFieldMatcher parseFieldMatcher) {
-        this.parseContext.parseFieldMatcher(parseFieldMatcher);
-    }
-
+    /**
+     * @return the parseFieldMatcher configured in the index settings
+     */
     public ParseFieldMatcher parseFieldMatcher() {
-        return parseContext.parseFieldMatcher();
+        return indexSettings.getParseFieldMatcher();
```

Comment:
So you suggest removing the parseFieldMatcher from IndexSettings and and instead hold an instance insice QueryShardContext instead? As far as I can see we need IndexSettings in QueryShardContext for the super ctor anyway, so my idea was to have it only in one place. Keeping the "real" instance in IndexSettings has the advantage that it only has to be creates once, while when moving it to QueryShardContext, we will have to create it in the ctor every time, but maybe I missunderstand what you mean.